File: Swift/프로그래머스/3/43162. 네트워크/네트워크.swift
import Foundation

var visited: [Bool] = []
var com: [[Int]] = []

func solution(_ n:Int, _ computers:[[Int]]) -> Int {
    var result: Int = 0
    com = computers
    visited = .init(repeating: false, count: n)
    
    for i in 0..<n {
        guard !visited[i] else { continue }
        dfs(i)
        result += 1
    }
    
    return result
}

func dfs(_ current: Int) {
    visited[current] = true
    
    let connecting = com[current]
    for i in 0..<connecting.count {
        guard i != current else { continue }
        guard connecting[i] == 1, !visited[i] else { continue }
        dfs(i)
    }
}

File: Java/프로그래머스/3/43162. 네트워크/네트워크.java
class Solution {
    static boolean[] visited;
    static int[][] com;
    
    public int solution(int n, int[][] computers) {
        int answer = 0;
        com = computers;
        visited = new boolean[n];
        
        for (int item = 0; item < n; item++) {
            if (!visited[item]) {
                dfs(item);
                answer++;
            }
        }
        
        return answer;
    }
    
    static void dfs(int current) {
        visited[current] = true;
        
        int[] connecting = com[current];
        for (int i = 0; i < connecting.length; i++) {
            if (i == current)
                continue;
            
            if (connecting[i] == 1 && !visited[i]) {
                dfs(i);
            }
        }
    }
}